api/v1/flow/flow.proto

```diff
@@ -375,6 +375,9 @@ message FlowFilter {
     repeated string http_method = 21;
     // http_path is a list of regular expressions to filter on the HTTP path.
     repeated string http_path = 22;
+
+    // tcp_flags filters flows based on TCP header flags
+    repeated TCPFlags tcp_flags = 23;
```

Comment:
Looks like these are the semantics of the `FlowFilter` type; all members are marked as `repeated` and, according to the type's description, are ANDed together and must all be true for a given flow to be emitted. That doesn't make sense in all cases, though, as eg. an HTTP request can only have a single method. :thinking: Could anyone shine some light on this?

Reply:
hi @kAworu 👋 

@ti-mo i'll give it a shot.

- repeated values for each individual filter get ORrd, so if you specify `http_method: ["PUT", "GET"]`, it means `PUT` or `GET`.
- then all the fields in the filter get ANDed, so `http_method: ["PUT", "GET"], http_path: ["/foo", "/bar"]`, it means `("PUT" or "GET") and ("/foo" or "/bar")`.
- then all the filters (filter itself can be repeated) get ORrd 🤯 

`TCPFlags` filter is an interesting case since the message itself contains all the flags unlike fields like `http_method` as @kAworu pointed out. personally i can go either way. using `repeated` looks more consistent with other fields, but having a single `TCPFlags` field might simplify the code a bit. @kAworu what do you think?

Reply:
@michi-covalent thanks a lot for the clarification, I think I got confused by this part of the commit message:

> In case multiple flags are used, the relational operator OR is used i.e. if any of the flags is set then the filter allows it.

Here is my understanding after a more careful read of the implementation:

when the filter is `tcp_flags: [[SYN, ACK]]`, then
1. flows with only the SYN flag will be _rejected_ by the filter
2. flows with only the ACK flag will be _rejected_ by the filter
1. flows with both the SYN and ACK flags will be _accepted_ by the filter

In other words we have a AND behaviour within a single `TCPFlags` field. There is no analog behaviour for `http_method` as HTTP methods are all mutually exclusive (as already correctly pointed out by @ti-mo) unlike TCP flags.

when the filter is `tcp_flags: [[SYN], [ACK]]`, then
1. flows with only the SYN flag will be _accepted_ by the filter
2. flows with only the ACK flag will be _accepted_ by the filter
1. flows with both the SYN and ACK flags will be _accepted_ by the filter

In other words we have the ORed repeated values for each individual filter that @michi-covalent described, which is consistent with other filters.

So repeated make sense to me now as it allow both behaviour whereas a single `TCPFlags` field would not. I can see a use-case for a filter like `tcp_flags: [[SYN, ACK], [ACK, FIN]]` which would be impossible to describe with a single `TCPFlags` field.

Reply:
> @michi-covalent thanks a lot for the clarification, it helped me understand that I got confused by this part of the commit message:
> 
> > In case multiple flags are used, the relational operator OR is used i.e. if any of the flags is set then the filter allows it.
> 
> Here is my understanding after a more careful read of the implementation:
> 
> when the filter is `tcp_flags: [[SYN, ACK]]`, then
> 
> 1. flows with only the SYN flag will be _rejected_ by the filter
> 2. flows with only the ACK flag will be _rejected_ by the filter
> 3. flows with both the SYN and ACK flags will be _accepted_ by the filter
> 
> In other words we have a AND behaviour within a single `TCPFlags` field. There is no analog behaviour for `http_method` as HTTP methods are all mutually exclusive (as already correctly pointed out by @ti-mo) unlike TCP flags.
> 
> when the filter is `tcp_flags: [[SYN], [ACK]]`, then
> 
> 1. flows with only the SYN flag will be _accepted_ by the filter
> 2. flows with only the ACK flag will be _accepted_ by the filter
> 3. flows with both the SYN and ACK flags will be _accepted_ by the filter

This is correct. The tests added to (tcp_test.go)[] also confirm this behaviour.

> 
> In other words we have the ORed repeated values for each individual filter that @michi-covalent described, which is consistent with other filters.
> 
> So "repeated" make sense to me now as it allow both behaviour whereas a single `TCPFlags` field would not. I can see a use-case for a filter like `tcp_flags: [[SYN, ACK], [ACK, FIN]]` which would be impossible to describe with a single `TCPFlags` field.



> @michi-covalent thanks a lot for the clarification, it helped me understand that I got confused by this part of the commit message:
> 
> > In case multiple flags are used, the relational operator OR is used i.e. if any of the flags is set then the filter allows it.
> 
> Here is my understanding after a more careful read of the implementation:
> 
> when the filter is `tcp_flags: [[SYN, ACK]]`, then
> 
> 1. flows with only the SYN flag will be _rejected_ by the filter
> 2. flows with only the ACK flag will be _rejected_ by the filter
> 3. flows with both the SYN and ACK flags will be _accepted_ by the filter

This is correct. The tests I added in `tcp_test.go` also confirms this behavior. My commit/PR message is incorrect.

> 
> In other words we have a AND behaviour within a single `TCPFlags` field. There is no analog behaviour for `http_method` as HTTP methods are all mutually exclusive (as already correctly pointed out by @ti-mo) unlike TCP flags.
> 
> when the filter is `tcp_flags: [[SYN], [ACK]]`, then
> 
> 1. flows with only the SYN flag will be _accepted_ by the filter
> 2. flows with only the ACK flag will be _accepted_ by the filter
> 3. flows with both the SYN and ACK flags will be _accepted_ by the filter
> 
> In other words we have the ORed repeated values for each individual filter that @michi-covalent described, which is consistent with other filters.
> 
> So "repeated" make sense to me now as it allow both behaviour whereas a single `TCPFlags` field would not. I can see a use-case for a filter like `tcp_flags: [[SYN, ACK], [ACK, FIN]]` which would be impossible to describe with a single `TCPFlags` field.

Reply:
The points discussed in this thread are correct. See previous discussion for the rationale here: https://github.com/cilium/cilium/pull/12932#discussion_r473977129